### ISSUE TITLE
Complaint because output file does not exist

### ISSUE THREAD

**Issue by sebersole** [eclipse/transformer#194]
I must be missing something simple about this tool and how it handles output dir/file.  First https://github.com/eclipse/transformer/issues/148

Now I run into an issue where the tool complains because the output file for a directory transformation does not exist.  Yes, the output file.  Isn't this what the tool is supposed to create?

Sorry if this is too much output.  Not really sure what is pertinent.

```
[main] INFO Transformer - Input     [ /home/sebersole/projects/hibernate-orm/main/hibernate-core-jakarta/target/tmp/unpack/test ]
[main] INFO Transformer -           [ /home/sebersole/projects/hibernate-orm/main/hibernate-core-jakarta/target/tmp/unpack/test ]
[main] INFO Transformer - Output    [ /home/sebersole/projects/hibernate-orm/main/hibernate-core-jakarta/target/tmp/transformed/test ] (Explicit)
[main] INFO Transformer -           [ /home/sebersole/projects/hibernate-orm/main/hibernate-core-jakarta/target/tmp/transformed/test ]
[main] INFO Transformer - Skipping option [ RULES_SELECTIONS ]
[main] INFO Transformer - External [ RULES_RENAMES ] URL [ file:/home/sebersole/projects/hibernate-orm/main/rules/jakarta-renames.properties ]
[main] INFO Transformer - External [ RULES_VERSIONS ] URL [ file:/home/sebersole/projects/hibernate-orm/main/rules/jakarta-versions.properties ]
[main] INFO Transformer - Internal [ RULES_BUNDLES ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-bundles.properties ]
[main] INFO Transformer - External [ RULES_DIRECT ] URL [ file:/home/sebersole/projects/hibernate-orm/main/rules/jakarta-direct.properties ]
[main] INFO Transformer - Internal [ RULES_MASTER_TEXT ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-text-master.properties ]
[main] INFO Transformer - Internal [ RULES_PER_CLASS_CONSTANT ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-per-class-constant-master.properties ]
[main] INFO Transformer - Package renames are in use
[main] INFO Transformer - Package versions will be updated
[main] INFO Transformer - Bundle identities will be updated
[main] INFO Transformer - Internal [ Substitions matching [ com.ibm.ws.transaction.services.TransactionSynchronizationRegistryService.xml ] ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-transactionSynchronizationRegistry.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ com.ibm.ws.transaction.services.UserTransactionService.xml ] ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-userTransaction.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ application-client.xml ] ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ TransactionJavaColonHelper.xml ] ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-userTransaction.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ com.ibm.ws.transaction.services.RemoteTransactionControllerService.xml ] ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-transactionManager.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ web.xml ] ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ com.ibm.ws.webcontainer.v40.dd.xml ] ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-servletVersion.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ UTService.xml ] ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-userTransaction.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ application.xml ] ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ TransactionManager.xml ] ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-transactionManager.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ ejb-jar.xml ] ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ web-fragment.xml ] ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ permissions.xml ] ] URL [ jar:file:/home/sebersole/.gradle/caches/modules-2/files-2.1/org.eclipse.transformer/org.eclipse.transformer.cli/0.2.0/bb8fcbee2323cb0a6078ed48955814fb48376c2a/org.eclipse.transformer.cli-0.2.0.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Text files will be updated
[main] INFO Transformer - Java direct string updates will be performed
[main] INFO Transformer - Per class constant mapping files are not enabled
[main] INFO Transformer - All resources will be selected
[main] INFO Transformer - Per class constant mapping files are not enabled
[main] INFO Transformer - Action selected for input [ /home/sebersole/projects/hibernate-orm/main/hibernate-core-jakarta/target/tmp/unpack/test ]: Directory Action
[main] ERROR Transformer - Transform failure:
org.eclipse.transformer.TransformException: Failed to open output [ /home/sebersole/projects/hibernate-orm/main/hibernate-core-jakarta/target/tmp/transformed/test/hibernate.properties ]
        at org.eclipse.transformer.action.impl.ActionImpl.openOutputStream(ActionImpl.java:656)
        at org.eclipse.transformer.action.impl.ActionImpl.apply(ActionImpl.java:620)
        at org.eclipse.transformer.action.impl.DirectoryActionImpl.transform(DirectoryActionImpl.java:103)
        at org.eclipse.transformer.action.impl.DirectoryActionImpl.transform(DirectoryActionImpl.java:93)
        at org.eclipse.transformer.action.impl.DirectoryActionImpl.apply(DirectoryActionImpl.java:63)
        at org.eclipse.transformer.Transformer$TransformOptions.transform(Transformer.java:1781)
        at org.eclipse.transformer.Transformer.run(Transformer.java:1874)
        at org.eclipse.transformer.jakarta.JakartaTransformer.main(JakartaTransformer.java:30)
Caused by: java.nio.file.NoSuchFileException: /home/sebersole/projects/hibernate-orm/main/hibernate-core-jakarta/target/tmp/transformed/test/hibernate.properties
        at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
        at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
        at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:219)
        at java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:478)
        at java.base/java.nio.file.Files.newOutputStream(Files.java:219)
        at aQute.lib.io.IO.outputStream(IO.java:1113)
        at aQute.lib.io.IO.outputStream(IO.java:1109)
        at org.eclipse.transformer.action.impl.ActionImpl.openOutputStream(ActionImpl.java:654)
        ... 7 more

```

**Comment by sebersole** [comment]
For what it is worth, I wondered if the problem is not what the tool actually complains about (the file not existing), but instead its real problem is the parent directory (not the target directory, see #148) not existing.

So I tried creating the parents of the target directory.  Now I get a slew of errors of the form:

```
[main] ERROR Transformer - Failed to parse raw class bytes [ ./test/org/hibernate/jpa/test/indetifier/DefaultInitialValueTableGeneratorConfiguredTest.class ]
java.io.IOException: Not a valid class file (no CAFEBABE header)
        at aQute.bnd.classfile.ClassFile.parseClassFile(ClassFile.java:69)
        at org.eclipse.transformer.action.impl.ClassActionImpl.apply(ClassActionImpl.java:291)
        at org.eclipse.transformer.action.impl.ActionImpl.basicApply(ActionImpl.java:592)
        at org.eclipse.transformer.action.impl.ActionImpl.apply(ActionImpl.java:570)
        at org.eclipse.transformer.action.impl.ActionImpl.apply(ActionImpl.java:622)
        at org.eclipse.transformer.action.impl.DirectoryActionImpl.transform(DirectoryActionImpl.java:103)
        at org.eclipse.transformer.action.impl.DirectoryActionImpl.transform(DirectoryActionImpl.java:93)
        at org.eclipse.transformer.action.impl.DirectoryActionImpl.transform(DirectoryActionImpl.java:93)
        at org.eclipse.transformer.action.impl.DirectoryActionImpl.transform(DirectoryActionImpl.java:93)
        at org.eclipse.transformer.action.impl.DirectoryActionImpl.transform(DirectoryActionImpl.java:93)
        at org.eclipse.transformer.action.impl.DirectoryActionImpl.transform(DirectoryActionImpl.java:93)
        at org.eclipse.transformer.action.impl.DirectoryActionImpl.transform(DirectoryActionImpl.java:93)
        at org.eclipse.transformer.action.impl.DirectoryActionImpl.apply(DirectoryActionImpl.java:63)
        at org.eclipse.transformer.Transformer$TransformOptions.transform(Transformer.java:1781)
        at org.eclipse.transformer.Transformer.run(Transformer.java:1874)
        at org.eclipse.transformer.jakarta.JakartaTransformer.main(JakartaTransformer.java:30)
```

Still no idea what the issue is here.  These are classes that were compiled and jar'ed in one project (hibernate-core) and then copied/unpacked in another (hibernate-core-jakarta).  It is at this point I try to apply the transformation but get this error.

**Comment by sebersole** [comment]
I should mention too that at this point MANY of the test classes have been transformed.  Nothing special about any of the ones it complains about

**Comment by bjhargrave** [comment]
I have a fix for creating the output file's parent directory.

As for the invalid class file issue, there was a problem with buffer growth when reading a file which was smaller than the buffer. This is now fixes in the latest 0.3.0-SNAPSHOT if you wish to check.
